test/parallel/test-internal-crypto-keys.js

```diff
@@ -0,0 +1,41 @@
+// Flags: --expose-internals
+'use strict';
+const common = require('../common');
+const keys = require('internal/crypto/keys');
+
+// This test ensures getKeyObjectHandle throws error for invalid key type
+{
+  const keyObject = new keys.KeyObject('private', {});
+  Object.defineProperty(keyObject, 'type', { get: () => 'test' });
+  common.expectsError(
+    () => {
+      keys.preparePublicOrPrivateKey(keyObject);
+    },
+    {
+      type: TypeError,
+      code: 'ERR_CRYPTO_INVALID_KEY_OBJECT_TYPE',
+      message: 'Invalid key object type test, expected private or public.'
+    }
+  );
+}
+
+// This test ensures KeyObject throws error for invalid key type
+// during the instantiation
+{
+  common.expectsError(() => new keys.KeyObject('someOther', {}), {
+    type: TypeError,
+    code: 'ERR_INVALID_ARG_VALUE',
+    message: "The argument 'type' is invalid. Received 'someOther'"
+  });
+}
+
+// This test ensures KeyObject throws error for invalid handler variable type
+// during the instantiation
+{
+  common.expectsError(() => new keys.KeyObject('private', 'notObject'), {
+    type: TypeError,
+    code: 'ERR_INVALID_ARG_TYPE',
+    message:
+      'The "handle" argument must be of type string. Received type string'
```

Comment:
Hmm... the error message here is not making sense.  `... must be of type string. Received type string` ...

It looks like this is an actual bug in the implementation at https://github.com/nodejs/node/blob/master/lib/internal/crypto/keys.js#L46-L47

```js
    if (typeof handle !== 'object')
      throw new ERR_INVALID_ARG_TYPE('handle', 'string', handle);
```

Should really be:

```js
    if (typeof handle !== 'object')
      throw new ERR_INVALID_ARG_TYPE('handle', 'Object', handle);
```

Also, I'm not seeing anything in the code that verifies that handle is not `null`, which would also be problematic.

That said, given that users really are not supposed to be creating `KeyObject` instances using the constructor, and we have validation logic that sits in front of these, these might be better modeled as `assert` statements rather than type validation errors.

Reply:
@jasnell , it make sense.

Reply:
@jasnell Regarding checking for `null`, I did not add such checks since the class wasn't meant to be exposed to users directly and no internal calls would ever pass `null`. I agree that the error handling can be improved.

Reply:
If null isn't reachable from the public API it should probably assert. And I'm not familiar enough with this code ATM to say for myself, but this really makes me wonder whether these tests against internals make sense (as I asked in https://github.com/nodejs/node/pull/27322#issuecomment-486706975). Testing internals can be useful to check conditions that are possible but hard to arrange with a unit test (like nework or protocol errors), or even perhaps to avoid having to create invalid DER objects just to trigger an error condition, but keyobjects should be entirely reachable through the public API, shouldn't they? And conditions that cannot be reached from the API don't need graceful error handling, asserts are fine.